application/src/main/java/org/thingsboard/server/service/entitiy/cf/DefaultTbCalculatedFieldService.java

```diff
@@ -15,58 +15,159 @@
-        this.calculatedFieldLinks = calculatedFieldService.findAllCalculatedFieldLinks().stream().collect(Collectors.toMap(CalculatedFieldLink::getId, cfl -> cfl));
-        this.calculatedFieldService = calculatedFieldService;
+    private ListeningExecutorService calculatedFieldExecutor;
+    private ListeningExecutorService calculatedFieldCallbackExecutor;
+
+    private final ConcurrentMap<CalculatedFieldId, CalculatedField> calculatedFields = new ConcurrentHashMap<>();
+    private final ConcurrentMap<CalculatedFieldId, List<CalculatedFieldLink>> calculatedFieldLinks = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, CalculatedFieldCtx> states = new ConcurrentHashMap<>();
+
+    @Value("${state.initFetchPackSize:50000}")
+    @Getter
+    private int initFetchPackSize;
+
+    @Value("10")
+    @Getter
+    private int defaultCalculatedFieldCheckIntervalInSec;
+
+    @PostConstruct
+    public void init() {
+        // from AbstractPartitionBasedService
+        scheduledExecutor = MoreExecutors.listeningDecorator(Executors.newSingleThreadScheduledExecutor(ThingsBoardThreadFactory.forName("calculated-field-scheduled")));
+        ///
+        calculatedFieldExecutor = MoreExecutors.listeningDecorator(ThingsBoardExecutors.newWorkStealingPool(
+                Math.max(4, Runtime.getRuntime().availableProcessors()), "calculated-field"));
+        calculatedFieldCallbackExecutor = MoreExecutors.listeningDecorator(ThingsBoardExecutors.newWorkStealingPool(
+                Math.max(4, Runtime.getRuntime().availableProcessors()), "calculated-field-callback"));
+        scheduledExecutor.scheduleWithFixedDelay(this::fetchCalculatedFields, new Random().nextInt(defaultCalculatedFieldCheckIntervalInSec), defaultCalculatedFieldCheckIntervalInSec, TimeUnit.SECONDS);
     }
 
-    @Override
-    public void onMsg() {
+    @PreDestroy
+    public void stop() {
+        // from AbstractPartitionBasedService
+        if (scheduledExecutor != null) {
+            scheduledExecutor.shutdown();
+        }
+        ///
+        if (calculatedFieldExecutor != null) {
+            calculatedFieldExecutor.shutdownNow();
+        }
+        if (calculatedFieldCallbackExecutor != null) {
+            calculatedFieldCallbackExecutor.shutdownNow();
+        }
+    }
 
+    @Override
+    public void onCalculatedFieldMsg(TransportProtos.CalculatedFieldMsgProto proto, TbCallback callback) {
+        try {
+            TenantId tenantId = TenantId.fromUUID(new UUID(proto.getTenantIdMSB(), proto.getTenantIdLSB()));
+            CalculatedFieldId calculatedFieldId = new CalculatedFieldId(new UUID(proto.getCalculatedFieldIdMSB(), proto.getCalculatedFieldIdLSB()));
+            if (proto.getDeleted()) {
+                onCalculatedFieldDelete(calculatedFieldId, callback);
+                callback.onSuccess();
+            }
+            if (proto.getUpdated()) {
+                onCalculatedFieldDelete(calculatedFieldId, callback);
```

Comment:
//TODO: improve the check. Maybe it was renamed or just the name of the result changed.